gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TheRestGroupRequestHandler.java

```diff
@@ -104,32 +103,124 @@ private ExploreRequest createRequest(
+   */
   private Filter.Builder createExcludeFoundGroupsFilter(
       ExploreRequest originalRequest, ExploreResponse.Builder originalResponse) {
+    if (originalRequest.getGroupByList().size() == 1) {
+      return createExcludeFoundGroupsNotInListFilter(originalRequest, originalResponse);
+    } else {
+      return createExcludeFoundGroupsAndChainFilter(originalRequest, originalResponse);
+    }
+  }
+
+  private Filter.Builder createExcludeFoundGroupsNotInListFilter(
+      ExploreRequest originalRequest, ExploreResponse.Builder originalResponse) {
     Filter.Builder filterBuilder = Filter.newBuilder();
     filterBuilder.setOperator(Operator.AND);
     originalRequest
         .getGroupByList()
         .forEach(
             groupBy -> {
               String columnName = groupBy.getColumnIdentifier().getColumnName();
-              Set<String> excludedValues = getExcludedValues(columnName, originalResponse);
+              List<String> excludedValues = getExcludedValues(columnName, originalResponse);
               filterBuilder.addChildFilter(createExcludedChildFilter(columnName, excludedValues));
             });
 
     return filterBuilder;
   }
 
-  private Set<String> getExcludedValues(
+  /**
+   * We need to create a filter that excludes all the group tuples found in the original response.
+   * Suppose we are dealing with a table where we group on the column names "c2" and "c3" and we got
+   * back the group tuples below in the original response.
+   * c2  | c3
+   * ---------
+   * v10 | v11
+   * v20 | v21
+   *
+   * So for "The Rest" request we need to exclude rows with the (c2,c3) tuples (v10, v11) or
+   * (v20, v21). So the filter should be ("=" is for equality):
+   *     ( NOT ( ( c2 = 'v10' AND c3 = 'v11' ) OR ( c2 = 'v20' AND c3 = 'v21' ) ) )
+   * However, since the query-service(Pinot) does not support NOT(unless it's NOT IN), then this does
+   * not work. So we can use DeMorgan's law to create a filter that Pinot will support by pushing
+   * the NOT into the filter expression:
+   *     ( ( ( c2 != 'v10' OR c3 != 'v11' ) AND ( c2 != 'v20' OR c3 != 'v21' ) ) )
+   * We will create the filter above.
+   * @param originalRequest
+   * @param originalResponse
+   * @return
+   */
+  private Filter.Builder createExcludeFoundGroupsAndChainFilter(
+      ExploreRequest originalRequest, ExploreResponse.Builder originalResponse) {
+    Filter.Builder filterBuilder = Filter.newBuilder();
+    filterBuilder.setOperator(Operator.AND);
+    List<String> groupByColumns = groupByColumnList(originalRequest);
+
+    originalResponse.getRowBuilderList().forEach(
+        rowBuilder -> filterBuilder.addChildFilter(createGroupValuesAndFilter(groupByColumns, rowBuilder))
+    );
+
+    return filterBuilder;
+  }
+
+  private List<String> groupByColumnList(ExploreRequest originalRequest) {
+    return originalRequest
+        .getGroupByList()
+        .stream()
+        .map(groupBy -> groupBy.getColumnIdentifier().getColumnName())
```

Comment:
Do we need a filter here `hasColumnIdentifier()`?

Reply:
No need. Only column names are allowed in the group by and the validators check this.